packages/runtime/src/ax.tsx

```diff
@@ -0,0 +1,36 @@
+/**
+ * Joins classes together and ensures atomic declarations of a single group exist.
+ * This function is made to be fast and small - so it takes some liberties with its API.
+ *
+ * Atomic declarations take the form of `_{group}{value}`,
+ * where both `group` and `value` are hashes **four characters long**.
+ *
+ * Input:
+ *
+ * ```
+ * ax(['_aaaabbbb', '_aaaacccc'])
+ * ```
+ *
+ * Output:
+ *
+ * ```
+ * '_aaaacccc'
+ * ```
+ *
+ * @param classes
+ */
+export const ax = (classes: (string | undefined | false)[]): string => {
+  const found: Record<string, string> = {};
```

Comment:
nit: Should we use `Map` instead of object? 🤔

Reply:
`Object.values(found).join(' ')` will become `found.values().join(' ')`

Reply:
let me run a quick test

Reply:
map values an iterable not an array

Reply:
map is much slower 

https://perf.link/#eyJpZCI6ImRka2pxM3B2Y2ZrIiwidGl0bGUiOiJGaW5kaW5nIG51bWJlcnMgaW4gYW4gYXJyYXkgb2YgMTAwMCIsImJlZm9yZSI6ImNvbnN0IGF4MSA9IChjbGFzc2VzKSA9PiB7XG4gIGNvbnN0IGZvdW5kID0ge307XG5cbiAgZm9yIChsZXQgaSA9IDA7IGkgPCBjbGFzc2VzLmxlbmd0aDsgaSsrKSB7XG4gICAgY29uc3QgY2xzID0gY2xhc3Nlc1tpXTtcbiAgICBpZiAoIWNscykge1xuICAgICAgY29udGludWU7XG4gICAgfVxuXG4gICAgY29uc3QgZ3JvdXAgPSBjbHMuc2xpY2UoMCwgY2xzLmNoYXJDb2RlQXQoMCkgPT09IDk1ID8gNSA6IHVuZGVmaW5lZCk7XG4gICAgZm91bmRbZ3JvdXBdID0gY2xzO1xuICB9XG5cbiAgcmV0dXJuIE9iamVjdC52YWx1ZXMoZm91bmQpLmpvaW4oJyAnKTtcbn07XG5cbmNvbnN0IGF4MiA9IChjbGFzc2VzKSA9PiB7XG4gIGNvbnN0IGZvdW5kID0gbmV3IE1hcCgpO1xuXG4gIGZvciAobGV0IGkgPSAwOyBpIDwgY2xhc3Nlcy5sZW5ndGg7IGkrKykge1xuICAgIGNvbnN0IGNscyA9IGNsYXNzZXNbaV07XG4gICAgaWYgKCFjbHMpIHtcbiAgICAgIGNvbnRpbnVlO1xuICAgIH1cblxuICAgIGNvbnN0IGdyb3VwID0gY2xzLnNsaWNlKDAsIGNscy5jaGFyQ29kZUF0KDApID09PSA5NSA%2FIDUgOiB1bmRlZmluZWQpO1xuICAgIGZvdW5kLnNldChncm91cCwgY2xzKTtcbiAgfVxuXG4gIHJldHVybiBBcnJheS5mcm9tKGZvdW5kLnZhbHVlcygpKS5qb2luKCcgJyk7XG59OyIsInRlc3RzIjpbeyJuYW1lIjoiVGVzdCBDYXNlIiwiY29kZSI6ImF4MShbJ2ZvbycsICdfMTIzMTIzJywgJ18xMjMxMjMnLCAnXzIzMjMyMzIzJ10pOyIsInJ1bnMiOlsxMDAwLDEwMDAsMTcwMDAsMjAwMCwxMDAwLDEzNDAwMCwxODAwMCwyMDAwLDMwMDAsMTAwMCw1OTAwMCwyMDAwLDIwMDAsMTU4MDAwLDMwODAwMCwyMTAwMCwxMDAwMCwyMDAwLDcwMDAsOTgwMDAsMTAwMCwwLDIwMDAsMTkyMDAwLDEwNzAwMCwyMDAwLDEwMDAsMjAwMCwyMDAwLDc4MDAwLDIwMDAsMjg2MDAwLDEwMDAwLDkwMDAsMTA5MDAwLDIyNDAwMCwxMDAwLDY2MDAwLDEwMDAwLDMwMDAsMzAwMCwzMDAwLDIwMDAsMzAwMCwxNjgwMDAsMzAwMCwzMDAwLDIwMDAsMzAwMCwzMTAwMCwzMDAwLDQwMDAsMTcwMDAsMzAwMCwzMDAwLDYwMDAsMjAwMCwyMDAwLDE2ODAwMCwxMDAwMCwyMDAwLDMwMDAsMTAwMDAsMjAwMCwwLDcxMDAwLDQwMDAsMzAwMCwzNDAwMCwyMDAwLDEwMDAwLDY3MDAwLDE2MDAwLDMwMDAsMjAwMCwyMDAwLDIwMDAsMjAwMCwzMDAwLDEwMDAsMjAwMCwzMDAwLDIwMDAsMjAwMCwyMDAwLDMwMDAsMjAwMCwyNTAwMCwyMDAwLDMwMDAsMzAwMCwzMDAwLDMwMDAsMjAwMCwyMDAwLDIwMDAsMzAwMCwxNTUwMDAsMTY4MDAwLDIwMDBdLCJvcHMiOjMwMjEwfSx7Im5hbWUiOiJUZXN0IENhc2UiLCJjb2RlIjoiYXgyKFsnZm9vJywgJ18xMjMxMjMnLCAnXzEyMzEyMycsICdfMjMyMzIzMjMnXSk7IiwicnVucyI6WzEwMDAsMTAwMCwzMDAwLDIwMDAsMTAwMCwxMDAwMCwzMDAwLDIwMDAsMzAwMCwxMDAwLDQwMDAsMjAwMCwyMDAwLDEwMDAwLDQyMDAwLDQwMDAsMzAwMCwyMDAwLDMwMDAsMTEwMDAsMTAwMCwwLDEwMDAsMjAwMDAsMTAwMDAsMjAwMCwxMDAwLDIwMDAsMjAwMCw1MDAwLDEwMDAsNTcwMDAsMzAwMCwzMDAwLDkwMDAsMjUwMDAsMTAwMCw0MDAwLDMwMDAsMjAwMCwzMDAwLDIwMDAsMjAwMCwyMDAwLDE2MDAwLDIwMDAsMzAwMCwyMDAwLDMwMDAsMzAwMCwzMDAwLDMwMDAsMzAwMCwyMDAwLDIwMDAsMzAwMCwyMDAwLDIwMDAsMTcwMDAsMzAwMCwyMDAwLDIwMDAsMzAwMCwyMDAwLDAsMzAwMCwzMDAwLDIwMDAsMzAwMCwxMDAwLDMwMDAsNTAwMCwzMDAwLDIwMDAsMjAwMCwyMDAwLDIwMDAsMjAwMCwyMDAwLDEwMDAsMTAwMCwyMDAwLDEwMDAsMjAwMCwyMDAwLDMwMDAsMTAwMCwzMDAwLDEwMDAsMzAwMCwyMDAwLDMwMDAsMzAwMCwxMDAwLDIwMDAsMTAwMCwzMDAwLDEyMDAwLDE3MDAwLDIwMDBdLCJvcHMiOjQ1MDB9XSwidXBkYXRlZCI6IjIwMjAtMDktMThUMDc6NDI6NDQuOTc5WiJ9

Reply:
Yeah 😅 hehe.`values()` will be an iterable. Forgot. hehe. But I just checked. If we remove `join('  ')` then `Map` is faster in most of the cases.

https://perf.link/#eyJpZCI6ImRka2pxM3B2Y2ZrIiwidGl0bGUiOiJGaW5kaW5nIG51bWJlcnMgaW4gYW4gYXJyYXkgb2YgMTAwMCIsImJlZm9yZSI6ImNvbnN0IGF4MSA9IChjbGFzc2VzKSA9PiB7XG4gIGNvbnN0IGZvdW5kID0ge307XG5cbiAgZm9yIChsZXQgaSA9IDA7IGkgPCBjbGFzc2VzLmxlbmd0aDsgaSsrKSB7XG4gICAgY29uc3QgY2xzID0gY2xhc3Nlc1tpXTtcbiAgICBpZiAoIWNscykge1xuICAgICAgY29udGludWU7XG4gICAgfVxuXG4gICAgY29uc3QgZ3JvdXAgPSBjbHMuc2xpY2UoMCwgY2xzLmNoYXJDb2RlQXQoMCkgPT09IDk1ID8gNSA6IHVuZGVmaW5lZCk7XG4gICAgZm91bmRbZ3JvdXBdID0gY2xzO1xuICB9XG5cbiAgcmV0dXJuIE9iamVjdC52YWx1ZXMoZm91bmQpO1xufTtcblxuY29uc3QgYXgyID0gKGNsYXNzZXMpID0%2BIHtcbiAgY29uc3QgZm91bmQgPSBuZXcgTWFwKCk7XG5cbiAgZm9yIChsZXQgaSA9IDA7IGkgPCBjbGFzc2VzLmxlbmd0aDsgaSsrKSB7XG4gICAgY29uc3QgY2xzID0gY2xhc3Nlc1tpXTtcbiAgICBpZiAoIWNscykge1xuICAgICAgY29udGludWU7XG4gICAgfVxuXG4gICAgY29uc3QgZ3JvdXAgPSBjbHMuc2xpY2UoMCwgY2xzLmNoYXJDb2RlQXQoMCkgPT09IDk1ID8gNSA6IHVuZGVmaW5lZCk7XG4gICAgZm91bmQuc2V0KGdyb3VwLCBjbHMpO1xuICB9XG5cbiAgcmV0dXJuIGZvdW5kLnZhbHVlcygpO1xufTsiLCJ0ZXN0cyI6W3sibmFtZSI6IlRlc3QgQ2FzZSIsImNvZGUiOiJheDEoWydmb28nLCAnXzEyMzEyMycsICdfMTIzMTIzJywgJ18yMzIzMjMyMyddKTsiLCJydW5zIjpbMjk3MDAwLDM2MjAwMCw0NDMwMDAsMzQ2MDAwLDIxNjAwMCwxNzQwMDAsNDA0MDAwLDIwNTAwMCw0ODkwMDAsMjc3MDAwLDIwMTAwMCwxOTQwMDAsNTk4MDAwLDg0MDAwLDQ5OTAwMCw5MDAwLDYyNjAwMCwxNjUwMDAsMTEwMDAsMTAwMCwzNjkwMDAsMTgwMDAwLDEwMDAsMTcxMDAwLDYwMzAwMCwzNDgwMDAsMzY5MDAwLDI0OTAwMCw3MDAwLDY1MDAwMCwyNTEwMDAsNjQzMDAwLDU3NzAwMCw0NzkwMDAsMzI4MDAwLDY4MDAwMCwyODYwMDAsNjAwMDAsMTM1MDAwLDQ0NzAwMCwxMzkwMDAsMjcwMDAsNDQwMDAwLDE5NzAwMCwxMDMwMDAsNDE2MDAwLDkzMDAwLDIzMDAwMCw1MjYwMDAsMzYzMDAwLDE2NzAwMCwyNTkwMDAsNDAxMDAwLDYxNDAwMCwyNjQwMDAsMzc1MDAwLDE0MjAwMCwxMzkwMDAsMTcwMDAsODgwMDAsNDY3MDAwLDQwNzAwMCwxNzgwMDAsMzQyMDAwLDM0MDAwLDQ4MzAwMCwyODUwMDAsMjE0MDAwLDExNDAwMCwyMDAwLDg0MDAwLDE5MTAwMCw0MTIwMDAsMjA1MDAwLDkwMDAwLDM1NjAwMCw1NDQwMDAsNzAzMDAwLDQ1ODAwMCwzODYwMDAsMTAwMCw0MDIwMDAsMzAwMCwzNzAwMCw2MDAwMCwxMzkwMDAsNDQ5MDAwLDI0OTAwMCw0MjIwMDAsNjk4MDAwLDQ0NzAwMCwyNzQwMDAsOTkwMDAsNTk4MDAwLDEwMDAsMTc3MDAwLDQ4ODAwMCw1NTUwMDAsNTQyMDAwLDI2NzAwMF0sIm9wcyI6MjkyOTcwfSx7Im5hbWUiOiJUZXN0IENhc2UiLCJjb2RlIjoiYXgyKFsnZm9vJywgJ18xMjMxMjMnLCAnXzEyMzEyMycsICdfMjMyMzIzMjMnXSk7IiwicnVucyI6WzE1MTAwMCwyNzAwMCwyNDgwMDAsMzU5MDAwLDI1OTAwMCw2NDAwMDAsMTAwMCw0MTAwMDAsNDkxMDAwLDUzODAwMCwzMTgwMDAsNDE4MDAwLDM3MDAwLDg0MDAwLDc0MDAwLDMwODAwMCw3MTIwMDAsMzUxMDAwLDc4MzAwMCw1MzgwMDAsNTk4MDAwLDE5ODAwMCwyMzQwMDAsODAzMDAwLDQ0NzAwMCwyOTUwMDAsMTkwMDAwLDU4MTAwMCw3MzMwMDAsMjA2MDAwLDY1NzAwMCw4MDAwLDUzNzAwMCw2NTAwMCwxNzgwMDAsMjA0MDAwLDQ1NDAwMCwzMDUwMDAsMjU2MDAwLDU0MTAwMCwyMTAwMCwzMDAwMCw0OTgwMDAsNjQxMDAwLDE4OTAwMCw1OTIwMDAsMzc4MDAwLDM3MTAwMCw0MjQwMDAsMTQ3MDAwLDM1OTAwMCw2NDgwMDAsMzk2MDAwLDY1NjAwMCw2NzcwMDAsMjk4MDAwLDE0NjAwMCw3MjQwMDAsMTAwMCwyODAwMCwxNjcwMDAsMTA4MDAwLDY1MTAwMCwzMjAwMCw1MzcwMDAsNTYxMDAwLDgwMDAsNTg2MDAwLDQ1NjAwMCw0ODgwMDAsNTA0MDAwLDc5MDAwLDU3MDAwLDE2MDAwMCw0OTAwMCwxNzgwMDAsMjU1MDAwLDU3MDAwMCwxMDcwMDAsNDQ0MDAwLDE1NDAwMCw0MDAwLDYzMjAwMCwzNDAwMCwzNDQwMDAsNDMwMDAsMzUwMDAwLDE2NzAwMCwyOTcwMDAsNzM2MDAwLDE4NDAwMCwxMjQwMDAsNTAwMCw2OTMwMDAsNjc3MDAwLDIyNjAwMCw2MTEwMDAsMjQ5MDAwLDkzMDAwLDY5OTAwMF0sIm9wcyI6MzM3ODAwfV0sInVwZGF0ZWQiOiIyMDIwLTA5LTE4VDA4OjA0OjEwLjQ1NVoifQ%3D%3D

Reply:
I think `Map` is only preferred when keys order matters. Coz it maintains the order of keys which is costly.

Reply:
If performance matters, probably object is preferred. `Map` will be slower because of maintaining insertion order (+ increase in space complexity) and `.get()` increases a function call.

Reply:
perf will matter here because its going to be a hot path

order only matters on input - output it can be in any order 😄

Reply:
will take any suggestions to improve perf!!

Reply:
hehe it's completely fine. If perf matters lets go with object. Coz insertion doesn't matter in this case. 🎉 😄

Reply:
it was just a **nit** in the first place though hehe 😛

Reply:
cool no wakkas